src/main.c

```diff
@@ -672,13 +674,22 @@ void handleApdu(unsigned int *flags, unsigned int *tx) {
                     break;
 
                 case INS_SIGN_EIP_712_MESSAGE:
-                    memset(tmpCtx.transactionContext.tokenSet, 0, MAX_ITEMS);
-                    handleSignEIP712Message(G_io_apdu_buffer[OFFSET_P1],
-                                            G_io_apdu_buffer[OFFSET_P2],
-                                            G_io_apdu_buffer + OFFSET_CDATA,
-                                            G_io_apdu_buffer[OFFSET_LC],
-                                            flags,
-                                            tx);
+                    if (G_io_apdu_buffer[OFFSET_P2] == 0) {
+                        memset(tmpCtx.transactionContext.tokenSet, 0, MAX_ITEMS);
+                        handleSignEIP712Message_v0(G_io_apdu_buffer[OFFSET_P1],
+                                                   G_io_apdu_buffer[OFFSET_P2],
+                                                   G_io_apdu_buffer + OFFSET_CDATA,
+                                                   G_io_apdu_buffer[OFFSET_LC],
+                                                   flags,
+                                                   tx);
+                    } else {
+#ifdef HAVE_EIP712_FULL_SUPPORT
+                        *flags |= IO_ASYNCH_REPLY;
+                        handle_eip712_sign(G_io_apdu_buffer);
+#else
+                        THROW(0x6B00);
```

Comment:
APDU_RESPONSE_INVALID_P1_P2

Reply:
Fixed in [4db1e99](https://github.com/LedgerHQ/app-ethereum/pull/327/commits/4db1e99fabbae80aa4288b035f4a012d5c6273f2)